virtcontainers/cgroups.go

```diff
@@ -1,197 +1,251 @@
+func (s *Sandbox) updateCgroups() error {
+	if s.state.CgroupPath == "" {
+		return fmt.Errorf("Cgroup path is empty")
 	}
 
-	resource, err := s.mergeSpecResource()
+	cgroup, err := cgroups.Load(V1Constraints, cgroups.StaticPath(s.state.CgroupPath))
 	if err != nil {
+		return fmt.Errorf("Could not load cgroup %v: %v", s.state.CgroupPath, err)
+	}
+
+	if err := s.constrainHypervisor(cgroup); err != nil {
 		return err
 	}
 
-	if err := s.applyCPUCgroup(resource); err != nil {
+	if len(s.containers) <= 1 {
+		// nothing to update
+		return nil
+	}
+
+	resources, err := s.resources()
+	if err != nil {
 		return err
 	}
+
+	if err := cgroup.Update(&resources); err != nil {
+		return fmt.Errorf("Could not update cgroup %v: %v", s.state.CgroupPath, err)
+	}
+
 	return nil
 }
 
-func (s *Sandbox) applyCPUCgroup(rc *specs.LinuxResources) error {
-	if s.cgroup == nil {
-		return fmt.Errorf("failed to setup uninitialized cgroup for sandbox")
+func (s *Sandbox) deleteCgroups() error {
+	path := cgroupNoConstraintsPath(s.state.CgroupPath)
+	noConstraintsCgroup, err := cgroups.Load(V1NoConstraints, cgroups.StaticPath(path))
+	if err != nil {
+		return fmt.Errorf("Could not load cgroup %v: %v", path, err)
 	}
 
-	// apply cpu constraint to vcpu cgroup
-	if err := s.cgroup.vcpuSub.Update(rc); err != nil {
-		return err
+	return noConstraintsCgroup.Delete()
+}
+
+func (s *Sandbox) constrainHypervisor(cgroup cgroups.Cgroup) error {
+	pid := s.hypervisor.pid()
+	if pid <= 0 {
+		return fmt.Errorf("Invalid hypervisor PID: %d", pid)
 	}
 
-	// when new container joins, new CPU could be hotplugged, so we
-	// have to query fresh vcpu info from hypervisor for every time.
-	tids, err := s.hypervisor.getThreadIDs()
+	resources := &specs.LinuxResources{}
+	path := cgroupNoConstraintsPath(s.state.CgroupPath)
+	noConstraintsCgroup, err := cgroups.New(V1NoConstraints, cgroups.StaticPath(path), resources)
 	if err != nil {
-		return fmt.Errorf("failed to get thread ids from hypervisor: %v", err)
-	}
-	if tids == nil {
-		// If there's no tid returned from the hypervisor, this is not
-		// a bug. It simply means there is nothing to constrain, hence
-		// let's return without any error from here.
-		return nil
+		return fmt.Errorf("Could not create cgroup %v: %v", path, err)
 	}
 
-	// use Add() to add vcpu thread to s.cgroup, it will write thread id to
-	// `cgroup.procs` which will move all threads in qemu process to this cgroup
-	// immediately as default behaviour.
-	if len(tids.vcpus) > 0 {
-		if err := s.cgroup.sandboxSub.Add(cgroups.Process{
-			Pid: tids.vcpus[0],
-		}); err != nil {
-			return err
-		}
+	if err := noConstraintsCgroup.Add(cgroups.Process{Pid: pid}); err != nil {
+		return fmt.Errorf("Could not add hypervisor PID %d to cgroup %v: %v", pid, path, err)
 	}
 
-	for _, i := range tids.vcpus {
-		if i <= 0 {
-			continue
-		}
+	// Add qemu into cgroup
+	return cgroup.Add(cgroups.Process{Pid: pid})
```

Comment:
`pid` is added both to `noConstraintsCgroup` and `cgroup`, in the end only the last `Add` can take effect.
So I think `noConstraintsCgroup` isn't necessary and can be removed.

By the way, I didn't see `noConstraintsCgroup` created on my disk, not sure which part I missed.

Reply:
> So I think noConstraintsCgroup isn't necessary and can be removed.

no, it's needed, since memory cgroup provided by k8s is too small (~50 mb) to run a VM, currently we don't see any failure because we are not honouring the `CgroupsPath` and our memory cgroup doesn't have any constraint

> By the way, I didn't see noConstraintsCgroup created on my disk, not sure which part I missed.

With this patch, only memory cgroup is part of `noConstraintsCgroup` and it's created under `kata` directory, for example `/sys/fs/cgroup/memory/kata/kubepods/burstable/poddbbadeac-2326-11e9-9b35-525400123456/crio-c44e31931c5b67260c977bfeb49d0fd6b2ebe0ff618e201115bf71d4eba9a29b`

Reply:
I get your point. This sounds reasonable.